packages/host/app/components/matrix/forgot-password.gts

```diff
@@ -0,0 +1,445 @@
+    return this.confirmPasswordError ? 'invalid' : 'initial';
+  }
+
+  @action
+  private sendEmailValidation() {
+    if (!this.email) {
+      throw new Error(
+        `bug: should never get here: reset password button disabled when no email`,
+      );
+    }
+
+    this.state = {
+      type: 'validateEmail',
+      email: this.email,
+      sendAttempt: 1,
+    };
+    this.sendEmailValidationTask.perform();
+  }
+
+  @action
+  private resendEmailValidation() {
+    if (this.state.type !== 'waitForEmailValidation') {
+      throw new Error(
+        `invalid state: cannot resendEmailValidation() in state ${this.state.type}`,
+      );
+    }
+
+    this.state = {
+      ...this.state,
+      sendAttempt: this.state.sendAttempt++,
+    };
+    this.sendEmailValidationTask.perform();
+  }
+
+  private sendEmailValidationTask = restartableTask(async () => {
+    if (
+      this.state.type === 'initial' ||
+      this.state.type === 'resetPassword' ||
+      this.state.type === 'resetPasswordSuccess'
+    ) {
+      throw new Error(
+        `invalid state: cannot sendEmailValidation() in state ${this.state.type}`,
+      );
+    }
+
+    try {
+      let clientSecret = uuidv4();
+      let { sid } = await this.matrixService.client.requestPasswordEmailToken(
+        this.state.email,
+        clientSecret,
+        this.state.sendAttempt,
+        window.location.href + `&clientSecret=${clientSecret}`,
+      );
+      this.state = {
+        ...this.state,
+        type: 'waitForEmailValidation',
+        clientSecret,
+        sid,
+      };
+    } catch (e: any) {
+      if (isMatrixError(e)) {
+        this.emailError = 'No account with the given email address exists';
+      }
+      if (this.state.type === 'validateEmail') {
+        this.state = {
+          type: 'initial',
+        };
+      }
+
+      throw e;
+    }
+  });
+
+  private resetPassword = restartableTask(async () => {
+    if (this.state.type !== 'resetPassword') {
+      throw new Error(
+        `invalid state: cannot resetPassword() in state ${this.state.type}`,
+      );
+    } else if (!this.password) {
+      throw new Error(
+        `bug: should never get here: reset password button disabled when no password`,
+      );
+    } else if (!this.args.resetPasswordParams) {
+      throw new Error(
+        `bug: should never get here: reset password params is required`,
+      );
+    }
+
+    try {
+      await this.matrixService.client.setPassword(
+        {
+          threepid_creds: {
+            sid: this.args.resetPasswordParams.sid,
+            client_secret: this.args.resetPasswordParams.clientSecret,
+          },
+          type: 'm.login.email.identity',
+        },
+        this.password,
+        true,
+      );
+      this.state = {
+        ...this.state,
+        type: 'resetPasswordSuccess',
+      };
+    } catch (e: any) {
+      if (isMatrixError(e)) {
+        this.error = 'Please check your email to validate reset password';
```

Comment:
I'm confused by this one - isn't the email where you validate your email address taking you to this place? If that's the case then why would you suggest the user to go validate it again?  

Also consider showing a general error if something unknown happens e.g. a network error.

Reply:
Nice catch! I forgot to remove this error message. This error message was for the previous flow where user can get to reset password page without verifying the email.